src/crystal/system/unix/file.cr

```diff
@@ -0,0 +1,236 @@
+require "c/sys/file"
+
+# :nodoc:
+module Crystal::System
+  module File
+    extend self
+
+    def open(filename, mode, perm)
+      oflag = open_flag(mode) | LibC::O_CLOEXEC
+
+      fd = LibC.open(filename.check_no_null_byte, oflag, perm)
+      if fd < 0
+        raise Errno.new("Error opening file '#{filename}' with mode '#{mode}'")
+      end
+      fd
+    end
+
+    private def open_flag(mode)
+      if mode.size == 0
+        raise "Invalid access mode #{mode}"
+      end
+
+      m = 0
+      o = 0
+      case mode[0]
+      when 'r'
+        m = LibC::O_RDONLY
+      when 'w'
+        m = LibC::O_WRONLY
+        o = LibC::O_CREAT | LibC::O_TRUNC
+      when 'a'
+        m = LibC::O_WRONLY
+        o = LibC::O_CREAT | LibC::O_APPEND
+      else
+        raise "Invalid access mode #{mode}"
+      end
+
+      case mode.size
+      when 1
+        # Nothing
+      when 2
+        case mode[1]
+        when '+'
+          m = LibC::O_RDWR
+        when 'b'
+          # Nothing
+        else
+          raise "Invalid access mode #{mode}"
+        end
+      else
+        raise "Invalid access mode #{mode}"
+      end
+
+      oflag = m | o
+    end
+
+    def mktemp(name, extension)
+      tmpdir = tempdir + ::File::SEPARATOR
+      path = "#{tmpdir}#{name}.XXXXXX#{extension}"
+
+      if extension
+        fd = LibC.mkstemps(path, extension.bytesize)
```

Comment:
1. Missing `String#check_no_null_byte`
2. mkstemps(3) reads: `Since it will be modified, template must not be a string constant, but should be declared as a character array.` That's not obvious from the function itself (At least to a non-C user), there should be atleast a comment noting that the string will be modified (Which isn't "possible" from Crystal), or it could be copied into a temporary `Bytes`.

Reply:
ditto